Add stdin option (fixes #598)

svenstaro opened the pull request:
This adds `--stdin` which allows you to specify that a file should be provided on stdin. Since we can't parse the filename in that case, we'll have to force the user to specify `-S` as well.

I handled both the legacy and modern ways for stdin and tested both.

I think me providing a dummy WorkItem called "stdin" in case of stdin is not too bad of a hack. I tried to make the best usage of what was already there in order to keep the patch as small as possible.

This also fixes #598 so that's nice.

johnkslang commented on StandAlone/StandAlone.cpp:
```diff
@@ -396,7 +397,10 @@ void ProcessArguments(std::vector<std::unique_ptr<glslang::TWorkItem>>& workItem
 
                     // handle --word style options
-                    if (lowerword == "auto-map-bindings" ||  // synonyms
+                    if (lowerword == "stdin") {
```
These are currently being done in alphabetical order. Should follow suit.

Also, is there a convention of just "--" alone meaning stdin?

johnkslang commented on StandAlone/StandAlone.cpp:
```diff
@@ -1231,6 +1278,8 @@ void usage()
            "                                       uint32_t array named <name>\n"
            "                                       initialized with the shader binary code.\n"
            "  --vn <name>                          synonym for --variable-name <name>\n"
+           "  --stdin                              Read from stdin instead of from a file. You'll have to provide\n"
```
Need alphabetical order again.

johnkslang replied:
Also, looks like it overflows the current 80 character limit the others are adhering to.

johnkslang commented:
This should appear as a single commit to glslang, which does not need a history of the mistakes made and corrected while developing a single increment change. (E.g., use `squash` and `push --force`.)

The precedent for testing bits is not strong, and evolving, so there is a mix, but I pointed out the preferred form a few times.

svenstaro commented:
I addressed all of your issues. Additionally, while both --stdin and - exist, I think in this case it's good to make it explicit and make people use --stdin because we also require them to provide -S in this case. I don't have a very strong opinion on this though so whatever you prefer.

johnkslang commented on StandAlone/StandAlone.cpp:
```diff
@@ -606,6 +610,10 @@ void ProcessArguments(std::vector<std::unique_ptr<glslang::TWorkItem>>& workItem
     }
 
+    // Make sure that -S is always specified if --stdin is specified
+    if ((Options & EOptionStdin) && !(shaderStageName != nullptr))
```
Better to say `shaderStageName == nullptr` than `!(shaderStageName != nullptr)`, yes?

svenstaro replied:
Uh, you're right, of course!

johnkslang commented:
My apologies for not submitting my previous review comments. (I'd thought the were all instant comments, but same came in under an unsubmitted review.)

svenstaro commented on StandAlone/StandAlone.cpp:
```diff
@@ -973,10 +1008,15 @@ int C_DECL main(int argc, char* argv[])
     }
 
-    if (workList.empty()) {
+    if (workList.empty() && !(Options & EOptionStdin)) {
```
Nope, that would change the logic. However, I'm going to make it `((Options & EOptionStdin) == 0)` instead.

svenstaro commented:
I addressed your further issues.
